viewer/packages/tools/src/Measurement/MeasurementTool.ts

```diff
@@ -0,0 +1,286 @@
+  private readonly _measurementLabel: MeasurementLabel;
+  private _lineGeometry: LineGeometry;
+  private _lineMesh: THREE.Mesh;
+  private _lineMaterial: LineMaterial;
+  private readonly _linePosition: Float32Array;
+  private readonly _raycaster: THREE.Raycaster;
+  private readonly _options: MeasurementOptions;
+  private _pointSize: number;
+  private _distanceValue: string;
+  private _distanceToCamera: number;
+
+  private readonly _handleonPointerClick = this.onPointerClick.bind(this);
+  private readonly _handleonPointerMove = this.onPointerMove.bind(this);
+
+  private readonly _lineOptions: MeasurementLineOptions = {
+    lineWidth: 0.002,
+    color: 0x00ffff
+  };
+
+  constructor(viewer: Cognite3DViewer, options?: MeasurementOptions) {
+    super();
+    this._viewer = viewer;
+    this._options = options ?? {};
+    this._measurementLabel = new MeasurementLabel(this._viewer);
+    this._linePosition = new Float32Array(6);
+    this._raycaster = new THREE.Raycaster();
+    this._pointSize = 0.02;
+    this._distanceValue = '';
+    this._distanceToCamera = 0;
+  }
+
+  /**
+   * Set input handling.
+   */
+  private setupEventHandling() {
+    this._viewer.on('click', this._handleonPointerClick);
+  }
+
+  /**
+   * Remove input handling
+   */
+  private removeEventHandling() {
+    this._viewer.off('click', this._handleonPointerClick);
+  }
+
+  /**
+   * Add an measurement.
+   */
+  add(): void {
+    this.setupEventHandling();
+  }
+
+  /**
+   * Remove measurement.
+   */
+  remove(): void {
+    //remove measurement label, clear all mesh, geometry & event handling
+    this._measurementLabel.remove();
+    this.clearLineObjects();
+    this.removeEventHandling();
+  }
+
+  /**
+   * Clear all line objects mesh, geometry & material
+   */
+  private clearLineObjects() {
+    if (this._lineMesh) {
+      this._lineMesh.geometry.dispose();
+      this._lineMesh = null;
+    }
+    if (this._lineGeometry) {
+      this._lineGeometry.dispose();
+      this._lineGeometry = null;
+    }
+    if (this._lineMaterial) {
+      this._lineMaterial.dispose();
+      this._lineMaterial = null;
+    }
+  }
+
+  private async onPointerClick(event: MouseEvent) {
+    const { offsetX, offsetY } = event;
+
+    const intersection = await this._viewer.getIntersectionFromPixel(offsetX, offsetY);
+
+    if (intersection) {
+      this.addSphere(intersection.point);
+
+      if (!this._lineMesh) {
+        this.startMeasurement(intersection);
+      } else {
+        this.endMeasurement(intersection.point);
+      }
+      this._viewer.requestRedraw();
+    }
+  }
+
+  /**
+   * Start the measurement.
+   * @param intersection Intersection Object containing point & camera distance.
+   */
+  private startMeasurement(intersection: Intersection) {
+    this._viewer.domElement.addEventListener('mousemove', this._handleonPointerMove);
+    this.startLine(intersection.point);
+    this._distanceToCamera = intersection.distanceToCamera;
+  }
```

Comment:
Coupling dom manipulation and computation makes this harder to right tests. Consider separating for easier testability (ref. you had a question on how to make this testable).